identity_iota_core/src/rebased/iota/package.rs

```diff
@@ -0,0 +1,226 @@
+
+use std::collections::HashMap;
+use std::sync::LazyLock;
+
+use anyhow::Context;
+use identity_iota_interaction::types::base_types::ObjectID;
+use serde::Deserialize;
+use serde::Deserializer;
+use serde::Serialize;
+use tokio::sync::RwLock;
+use tokio::sync::RwLockReadGuard;
+use tokio::sync::RwLockWriteGuard;
+use tokio::sync::TryLockError;
+
+use crate::rebased::client::CoreClientReadOnly;
+use crate::rebased::Error;
+
+pub(crate) const MAINNET_CHAIN_ID: &str = "6364aad5";
+
+type PackageRegistryLock = RwLockReadGuard<'static, PackageRegistry>;
+type PackageRegistryLockMut = RwLockWriteGuard<'static, PackageRegistry>;
+
+static IDENTITY_PACKAGE_REGISTRY: LazyLock<RwLock<PackageRegistry>> = LazyLock::new(|| {
+  let move_lock_content = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/packages/iota_identity/Move.lock"));
+  RwLock::new(PackageRegistry::from_move_lock_content(move_lock_content).expect("Move.lock exists and it's valid"))
+});
+
+pub(crate) async fn identity_package_registry() -> PackageRegistryLock {
+  IDENTITY_PACKAGE_REGISTRY.read().await
+}
+
+pub(crate) fn try_identity_package_registry() -> Result<PackageRegistryLock, TryLockError> {
+  IDENTITY_PACKAGE_REGISTRY.try_read()
+}
+
+pub(crate) fn blocking_identity_registry() -> PackageRegistryLock {
+  IDENTITY_PACKAGE_REGISTRY.blocking_read()
+}
+
+pub(crate) async fn identity_package_registry_mut() -> PackageRegistryLockMut {
+  IDENTITY_PACKAGE_REGISTRY.write().await
+}
+
+pub(crate) fn try_identity_package_registry_mut() -> Result<PackageRegistryLockMut, TryLockError> {
+  IDENTITY_PACKAGE_REGISTRY.try_write()
+}
+
+pub(crate) fn blocking_identity_registry_mut() -> PackageRegistryLockMut {
+  IDENTITY_PACKAGE_REGISTRY.blocking_write()
+}
+
+/// Network / Chain information.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub(crate) struct Env {
+  pub chain_id: String,
+  pub alias: Option<String>,
+}
+
+impl Env {
+  /// Creates a new package's environment.
+  pub(crate) fn new(chain_id: impl Into<String>) -> Self {
+    Self {
+      chain_id: chain_id.into(),
+      alias: None,
+    }
+  }
+
+  /// Creates a new package's environment with the given alias.
+  pub(crate) fn new_with_alias(chain_id: impl Into<String>, alias: impl Into<String>) -> Self {
+    Self {
+      chain_id: chain_id.into(),
+      alias: Some(alias.into()),
+    }
+  }
+}
+
+/// A published package's metadata for a certain environment.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub(crate) struct Metadata {
+  pub original_published_id: ObjectID,
+  pub latest_published_id: ObjectID,
+  #[serde(deserialize_with = "deserialize_u64_from_str")]
+  pub published_version: u64,
+}
+
+impl Metadata {
+  /// Create a new [Metadata] assuming a newly published package.
+  pub(crate) fn from_package_id(package: ObjectID) -> Self {
+    Self {
+      original_published_id: package,
+      latest_published_id: package,
+      published_version: 1,
```

Comment:
Just wondering if we should set this to 1 here.

We don't seem to be using the version at the moment, but I'm guessing it is parsed and might be used in the future, so we parse it from the `.lock` file.

`IdentityClientReadOnly::new_with_pkg_id` uses the current function (`from_package_id`) to allow passing a package into the client, but then it is always set to 1. Was just wondering if we currently have the two possible setups:

- we parsed the `.lock` file and `published_version` is valid
- we used `new_with_pkg_id` and it is set to 1 and might be invalid

So I was thinking, I'm either missing an update for the 1, are planning to add this in a follow-up issue, or if we should make `published_version` optional and just set it to `None` here.